freqtrade/optimize/hyperopt_loss_sortino_daily.py

```diff
@@ -0,0 +1,69 @@
+"""
+SortinoHyperOptLossDaily
+
+This module defines the alternative HyperOptLoss class which can be used for
+Hyperoptimization.
+"""
+import math
+import statistics
+from datetime import datetime
+
+from pandas import DataFrame, date_range
+
+from freqtrade.optimize.hyperopt import IHyperOptLoss
+
+
+class SortinoHyperOptLossDaily(IHyperOptLoss):
+    """
+    Defines the loss function for hyperopt.
+
+    This implementation uses the Sortino Ratio calculation.
+    """
+
+    @staticmethod
+    def hyperopt_loss_function(results: DataFrame, trade_count: int,
+                               min_date: datetime, max_date: datetime,
+                               *args, **kwargs) -> float:
+        """
+        Objective function, returns smaller number for more optimal results.
+
+        Uses Sortino Ratio calculation.
+
+        Sortino Ratio calculated as described in
+        http://www.redrockcapital.com/Sortino__A__Sharper__Ratio_Red_Rock_Capital.pdf
+        """
+        resample_freq = '1D'
+        slippage_per_trade_ratio = 0.0005
+        days_in_year = 365
+        minimum_acceptable_return = 0.0
+
+        # apply slippage per trade to profit_percent
+        results.loc[:, 'profit_percent_after_slippage'] = \
+            results['profit_percent'] - slippage_per_trade_ratio
+
+        # create the index within the min_date and end max_date
+        t_index = date_range(start=min_date, end=max_date, freq=resample_freq,
+                             normalize=True)
+
+        sum_daily = (
+            results.resample(resample_freq, on='close_time').agg(
+                {"profit_percent_after_slippage": sum}).reindex(t_index).fillna(0)
+        )
+
+        total_profit = sum_daily["profit_percent_after_slippage"] - minimum_acceptable_return
+        expected_returns_mean = total_profit.mean()
+
+        sum_daily['downside_returns'] = 0
+        sum_daily.loc[total_profit < 0, 'downside_returns'] = total_profit
```

Comment:
```suggestion
        sum_daily.loc[total_profit < minimum_acceptable_return, 'downside_returns'] = total_profit
```

Reply:
let me think about proposed changes...

Reply:
if you don't make these changes, when the MAR is not 0, the calculation is wrong because if you apply the std dev function with the MAR as mean unit then you have to normalize the positive returns to the MAR, otherwise the std dev function will subtract the MAR to the points nulled before.
- either you ignore the MAR and just use 0 everywhere or normalize as suggested or calculate the downside deviation without the pstdev function

Reply:
I used this article as the reference for implementation:
http://www.redrockcapital.com/Sortino__A__Sharper__Ratio_Red_Rock_Capital.pdf

* The formula in wikipedia is a continuous one (with a random var and an integral) so it's useless for practical calculations;
* The formulas and examples of calculations in programmingforfinance, investopedia and xplaind (see at the top) are simplified;
* So this article seems to be the most "scientific-like".

If you take a look at the article, it's not the stdev at the denominator, because stdev is calculated as the second moment (well, sqrt of) against the mean, while the formula in the article is actually is a moment against the zero point (since MAR was already subtracted under the min() function). Please review my explainations/annotations above: https://github.com/freqtrade/freqtrade/pull/2879#issuecomment-587468323

Reply:
But anyway, I'll also review this once again...

Reply:
I suggested the modifications based on that article that you took as reference not on the formula with the integral

Reply:
@hroff-1902 I actually think orehunt's suggestions make sense, lets commit them and I'll test them and compare with the previous version using a randomstate I've run before.